Issue1940 dev docu

milicag opened the pull request:
Closes #1940. Integrates the style guide wiki page and aims to replace it. Integrates the PR checklist created by an array of control sequence developers.

milicag commented:
@mwetter The test fails on the Sphinx version. Please comment on the the edits to the content when you get a chance and I will use the aligned version to recreate the `html`s.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
+   differentiation. For example, in Dymola, if your model translation
+   log shows
+
+   ::
```
This should be rendered as ..code::

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
+
+   (or any number other than zero), then enter on the command line
+
+   ::
```
this should be rendered as ..code::

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
+
+Functions
+~~~~~~~~~
+
```
One blank line suffices.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
 
-#. Each block must have a ``defaultComponentName`` annotation.
+#. In the code text the instances must be ordered as follows:
+     - If an instance has a type associated with it, then list them in this order: `boolean`, `integer`, `real`.
```
I don't understand this. Every instance has a type. Also, use the Modelica types, e.g., `Real` and not `real`.

milicag replied:
An instance of a subsequence block does not have a type.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
+     - If an instance has a type associated with it, then list them in this order: `boolean`, `integer`, `real`.
+     - Instances of lower dimensionality should come first, e.g. scalar values before arrays, 
+       though grouping based on model specific similarities may be prioritized
+     - Instances should be ordered as: parameters, interfaces: first inputs, then outputs, then ports,
```
There are no `ports` in CDL  and simply call them inputs and outputs, not interfaces as interface is not defined what it is. Also, correct punctuation to use a full sentence.

milicag replied:
`ports` is legacy from previous version of the document, as it is for both CDL and other development within the library.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
+       followed by any blocks/instances
+     - Protected instances are below all the unprotected instances and follow the same instance ordering rules.
+
+#. Each block must have a ``defaultComponentName`` annotation and a ``%name`` at icon layer
```
Replace with "and a ``%name`` label placed above the icon.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
 #. For parameters, where generally valid values can be provided, provide them
    as default values.
 
+#. Note that the protected instances are not accessible with the plots used in the validation tests.
```
This does not belong here; let's not mix how tools work with how to implement sequences.

milicag replied:
Is there a better spot for it? Maybe in the testing section?

milicag replied:
I moved it to the end of the file.

milicag replied:
Moved to unit test wiki page.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
 
+#. Note that the protected instances are not accessible with the plots used in the validation tests.
+
+#. Add comments to all the instances such as parameters, interfaces, blocks. The comments should be short and concise. The comments 
```
There are no interfaces, and "concise" is by definition short.

milicag replied:
Agreed

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
+#. Add comments to all the instances such as parameters, interfaces, blocks. The comments should be short and concise. The comments 
+   should not contain redundant information and hard coded parameters. If the functionality of an instance is obvious the developer may use
+   comments that closely resemble the class names, such as `Logical And`. 
+   Make sure comments are short and don’t contain any redundant information.
```
This is said on line 563 already.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -219,25 +608,29 @@ The following rules need to be followed, in addition to the guidelines described
     Buildings.Controls.OBC.CDL.Interfaces.RealInput TZon(
       final unit="K",
-      displayUnit="degC") "Measured zone air temperature";
+      final displayUnit="degC") "Measured zone air temperature";
```
Display unit must not be final. Users may want to change this to IP.

milicag replied:
That makes perfect sense - however we've been using final with the displayUnit throughout the OBC package. It should be easy to replace.

milicag replied:
#2045

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -219,25 +608,29 @@ The following rules need to be followed, in addition to the guidelines described
    process CDL.
 
+#. Units, quantities and value limits must be declared as ``final`` to avoid users to be able to change them, as
+   a change in unit may cause the control logic to be incorrect. Often display units are also declared as final.
```
remove " Often display units are also declared as final."

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
    Make sure sequences are tested for all modes of operation, and as applicable, for winter, shoulder
    and summer days.
 
+#. Perform code text and, if applicable, figure inspection to:
```
Remove to be concise. It should be clear that code and figures need to be inspected.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
+#. Use full paths to blocks, that is paths that start with `Buildings.Controls...`
 
-.. _sec_val:
+#. Final steps for a pull requests:
```
Remove. This belongs to how to make a pull request. It is not specific to control sequences.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
-----------
+#. For general rules on validation models see :numref:`sec_val`. Specifically for the 
+   sequence development ensure that:
+      - the instances of models being tested are listed at the top of the code text
```
I don't understand this line.

milicag replied:
If I am testing a MZ VAV controller then an instance of it should come before any other block instances included in the test.

milicag replied:
It is our own recommendation, similarly to ordering of types or instances.

milicag replied:
I rephrased, see if it works

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
+#. For general rules on validation models see :numref:`sec_val`. Specifically for the 
+   sequence development ensure that:
+      - the instances of models being tested are listed at the top of the code text
+      - add a blank line at the end of each .mos script
```
Why is a blank line needed? If it is needed, is it a bug in BuildingsPy? If it is needed for Dymola, it belongs to the description of how to implement unit tests. This needs to be about control sequences only. If we start mixing it, we will repeat ourself and won't be concise.

milicag replied:
I assume it was added due to some tests failing if there isn't a blank line at the end of a mos file.

milicag replied:
I'll move these two statements to the unit testing wiki page:

* Protected instances are not accessible through `mos` scripts. 
* A blank line at the end of each .mos script

What's confusing is that we have a Validation and Unit Tests section in this file too.

milicag replied:
Moved.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
-realistic operating modes.
-These validations need to be part of the
-`unit tests <https://github.com/lbl-srg/modelica-buildings/wiki/Unit-Tests>`_.
+#. Run modelica-json$ node app.js -f Buildings/Controls/OBC/ASHRAE/PrimarySystem/{path to package} -o json -m modelica 
```
Format as code.

mwetter commented on Buildings/Resources/Documentation/userGuide/source/development.rst:
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
+   Unit tests should cover all branches of ``if-then`` constructs and 
+   all realistic operating modes of the system represented by the model.
+   See `unit test implementation <Unit-Tests>`__.
+2. Ensure that no unit test requires a numerical Jacobian. If a
```
This is already explained on line 187, remove it here.

mwetter commented on Buildings/Resources/Documentation/userGuide/build/html/workArounds.html:
```diff
@@ -322,7 +329,7 @@ <h2>3.5. Avoiding overspecified initialization problems<a class="headerlink" hre
     <p>
         &copy; Copyright 2012-2020, The Regents of the University of California (through Lawrence Berkeley National Laboratory).<br/>
-      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.<br/>
+      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.<br/>
```
It looks like you have an old version of sphinx.

milicag replied:
planning to update

milicag commented:
@mwetter I've addressed your comments and updated my Sphinx. 5.4.3 under 3. and 17. are formatted bold in the html and I don't see directly why. Please take a look again and let me know of any remaining issues.

mwetter commented:
@milicag : Please wait with further edits as I am reviewing it now.

mwetter commented:
@milicag : Do we really need these two statements below? I think the first is not needed because a Modelica tool takes care of this, and `modelica-json` should as well, and I am not sure whether the second is really sufficiently important.
```
#. Use full paths to blocks, that is paths that start with ``Buildings.Controls...``.
```

```
#. For general rules on validation models see :numref:`sec_val`. Preferably,
   list the block being tested at the top of the code, and list inputs to the block afterwards.
```
Once these are addressed and all tests pass, it is ready to merge. Then we should also delete https://github.com/lbl-srg/modelica-buildings/wiki/Style-Guide. I made the changes also to https://github.com/ibpsa/modelica-ibpsa/wiki/Style-Guide so that they are consistent.

milicag commented:
@mwetter 

first: If there is really no concern about the first one I'm happy to delete it.

second: it really isn't a must, I got used to that structure similarly to say order of types: Boolean, Integer, Real. At the very least I feel that the tested instances should be grouped together, such that we don't have one tested instance, then a bunch of say pulse signals, than a constant, then again tested instances. Up to you.

milicag commented on Buildings/Resources/Documentation/userGuide/build/html/_sources/development.rst.txt:
```diff
@@ -619,7 +619,7 @@ The following rules need to be followed, in addition to the guidelines described
 #. The size of the icon should be such that it provides a good fit for all the input and output interfaces. The minimum
    recommended icon size is 100 by a 100. If there are many interfaces the icon size should be extended in vertical direction.
-   Icons should be symetrical with reference to the grid origin. E.g, the default specification is
+   Icons should be symmetrical with reference to the grid origin. E.g, the default specification is
```
ups, I forgot the spell check isn't active as I've changed text editors. Apologies!
